pkg/auth/providers/keycloakoidc/keycloak_client.go

```diff
@@ -0,0 +1,113 @@
+package keycloakoidc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+)
+
+type searchResults struct {
+	Items []Account `json:"items"`
+}
+
+//Account defines properties an account on github has
+type Account struct {
+	ID       int      `json:"id,omitempty"`
+	Username string   `json:"username,omitempty"`
+	Enabled  bool     `json:"enabled,omitempty"`
+	Name     string   `json:"firstName,omitempty"`
+	LastName string   `json:"lastName,omitempty"`
+	Groups   []string `json:"groups"`
+}
+
+//KClient implements a httpclient for keycloak
+type KClient struct {
+	httpClient *http.Client
+}
+
+func (k *KClient) keycloakSearch(searchTerm, searchType string, accessToken string, config *v32.OIDCConfig) ([]Account, error) {
+	searchUrl := fmt.Sprintf("%s/%s?search=%s", k.getSearchUrl(config.Issuer), searchType, searchTerm)
+	search := URLEncoded(searchUrl)
+
+	b, err := k.getFromKeyCloak(accessToken, search)
+	if err != nil {
+		// no match on search returns an error. do not log
+		return nil, nil
+	}
+	var searchResults []Account
+	if err := json.Unmarshal(b, &searchResults); err != nil {
+		logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+		return searchResults, err
+	}
+	return searchResults, nil
+}
+
+func (k *KClient) getFromKeyCloakById(principalID, accessToken, searchType string, config *v32.OIDCConfig) (Account, error) {
+	searchUrl := fmt.Sprintf("%s/%s/%s", k.getSearchUrl(config.Issuer), searchType, principalID)
+	search := URLEncoded(searchUrl)
+	b, err := k.getFromKeyCloak(accessToken, search)
+	if err != nil {
+		return Account{}, nil
+	}
+	var searchResult Account
+	if err := json.Unmarshal(b, &searchResult); err != nil {
+		logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+		return searchResult, err
+	}
+	return searchResult, nil
+
+}
+func (k *KClient) getSearchUrl(issuer string) string {
+	splitIssuer := strings.SplitAfter(issuer, "/auth/")
+	return fmt.Sprintf(
+		"%sadmin/%s",
+		splitIssuer[0],
+		splitIssuer[1],
+	)
+}
+
+//URLEncoded encodes the string
+func URLEncoded(str string) string {
+	u, err := url.Parse(str)
+	if err != nil {
+		logrus.Errorf("Error encoding the url: %s, error: %v", str, err)
+		return str
+	}
+	return u.String()
+}
+
+func (k *KClient) getFromKeyCloak(accessToken string, url string) ([]byte, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "token "+accessToken)
```

Comment:
We are doing a GET request with an access token ? Can you not do a POST ? Or is the accessToken not sensitive ?

Reply:
Why would a POST be used to GET data? A POST doesn't hide the access token

Reply:
It does if it's in the body. I suppose this depends how keycloak can receive it.